packages/connection/src/class-error-handler.php

```diff
@@ -0,0 +1,569 @@
+		'could_not_sign',
+		'invalid_scheme',
+		'invalid_secret',
+		'invalid_token',
+		'token_mismatch',
+		'invalid_body',
+		'invalid_signature',
+		'invalid_body_hash',
+		'invalid_nonce',
+		'signature_mismatch',
+	);
+
+	/**
+	 * Holds the instance of this singleton class
+	 *
+	 * @since 8.7.0
+	 *
+	 * @var Error_Handler $instance
+	 */
+	public static $instance = null;
+
+	/**
+	 * Initialize instance, hookds and load verified errors handlers
+	 *
+	 * @since 8.7.0
+	 */
+	private function __construct() {
+		defined( 'JETPACK__ERRORS_PUBLIC_KEY' ) || define( 'JETPACK__ERRORS_PUBLIC_KEY', 'KdZY80axKX+nWzfrOcizf0jqiFHnrWCl9X8yuaClKgM=' );
+
+		add_action( 'rest_api_init', array( $this, 'register_verify_error_endpoint' ) );
+
+		$this->handle_verified_errors();
+
+		// If the site gets reconnected, clear errors.
+		add_action( 'jetpack_site_registered', array( $this, 'delete_all_errors' ) );
+	}
+
+	/**
+	 * Gets the list of verified errors and act upon them
+	 *
+	 * @since 8.7.0
+	 *
+	 * @return void
+	 */
+	public function handle_verified_errors() {
+		$verified_errors = $this->get_verified_errors();
+		foreach ( $verified_errors as $error_code => $user_errors ) {
+
+			switch ( $error_code ) {
+				case 'malformed_token':
+				case 'token_malformed':
+				case 'no_possible_tokens':
+				case 'no_valid_token':
+				case 'unknown_token':
+				case 'could_not_sign':
+				case 'invalid_token':
+				case 'token_mismatch':
+				case 'invalid_signature':
+				case 'signature_mismatch':
+					new Error_Handlers\Invalid_Blog_Token( $user_errors );
+					break;
+			}
+		}
+	}
+
+	/**
+	 * Gets the instance of this singleton class
+	 *
+	 * @since 8.7.0
+	 *
+	 * @return Error_Handler $instance
+	 */
+	public static function get_instance() {
+		if ( is_null( self::$instance ) ) {
+			self::$instance = new self();
+		}
+		return self::$instance;
+	}
+
+	/**
+	 * Keep track of a connection error that was encountered
+	 *
+	 * @since 8.7.0
+	 *
+	 * @param \WP_Error $error the error object.
+	 * @param boolean   $force Force the report, even if should_report_error is false.
+	 * @return void
+	 */
+	public function report_error( \WP_Error $error, $force = false ) {
+		if ( in_array( $error->get_error_code(), $this->known_errors, true ) && $this->should_report_error( $error ) || $force ) {
+			$stored_error = $this->store_error( $error );
+			if ( $stored_error ) {
+				$this->send_error_to_wpcom( $stored_error );
+			}
+		}
+	}
+
+	/**
+	 * Checks the status of the gate
+	 *
+	 * This protects the site (and WPCOM) against over loads.
+	 *
+	 * @since 8.7.0
+	 *
+	 * @param \WP_Error $error the error object.
+	 * @return boolean $should_report True if gate is open and the error should be reported.
+	 */
+	public function should_report_error( \WP_Error $error ) {
+
+		if ( defined( 'JETPACK_DEV_DEBUG' ) && JETPACK_DEV_DEBUG ) {
+			return true;
+		}
+
+		$bypass_gate = apply_filters( 'jetpack_connection_bypass_error_reporting_gate', false );
```

Comment:
Need a docblock on this

Reply:
done in https://github.com/Automattic/jetpack/pull/16194/commits/f02d8c78ba0f1cec4943f88770bb5bde46dec5ce